modules/civicrmtheme/src/Theme/CivicrmThemeNegotiator.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Drupal\civicrmtheme\Theme;
+
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\Theme\ThemeNegotiatorInterface;
+
+/**
+ * Theme negotiator for CiviCRM pages.
+ */
+class CivicrmThemeNegotiator implements ThemeNegotiatorInterface {
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $user;
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * Constructs a CivicrmThemeNegotiator.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The current user service.
+   */
+  public function __construct(AccountInterface $user, ConfigFactoryInterface $config_factory) {
+    $this->user = $user;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applies(RouteMatchInterface $route_match) {
+    $route = $route_match->getRouteObject();
+
+    // Some pages, like 404 pages, don't have a route objet.
+    if (!$route) {
+      return FALSE;
+    }
+
+    $parts = explode('/', ltrim($route->getPath(), '/'));
+
+    if ($parts[0] != 'civicrm') {
+      return FALSE;
+    }
+
+    if (count($parts) > 1 && $parts[1] == 'upgrade') {
+      return FALSE;
+    }
+
+    $config = $this->configFactory->get('civicrmtheme.settings');
+    $admin_theme = $config->get('admin_theme');
+    $public_theme = $config->get('public_theme');
+
+    // @todo Should we inject the 'civicrm' service instead?
+    if ((!$admin_theme && !$public_theme) || !\Drupal::service('civicrm')) {
```

Comment:
Here, should we be using the new initialize() method of the civicrm service instead, after that other PR gets merged?

Reply:
Or maybe not...you comment there though, "Should we inject the 'civicrm' service instead....Isn't dependency injection the standard Drupal 8 practice? I'm not criticizing, I'm asking he question....

Can you help me understand what that line is doing? line 66

Reply:
Basically the question is, for me, what do you need to know with that line?  If it can't load the service, then return false, meaning civicrm couldn't be initialized? Because the constructor would return the object....so if you get no object then return false?

Reply:
Yeah, you're right, that bit is not totally useful. I ported this module by starting with the D7 code and changing it until it was a working D8 module, and so, I was converting the `civicrm_initialize()` to `\Drupal::service('civicrm')` and ended up with that bit of nonsense. :-) The service cannot be falsey (without changing how the container creates it), it can only be the service object or throw an exception.

> Basically the question is, for me, what do you need to know with that line?

So, I ported this from the D7 code, but I believe that bit is about not trying to do anything if CiviCRM isn't working, just bail early.

I've made a new commit that makes a lot more sense!

We definitely shouldn't inject the 'civicrm' service until after #495 is merged, because then civicrm will be initialized on every request, not just those that starting with `/civicrm`, which could be a performance issue. But once that is merged, it would be best practice to inject it!